### PR TITLE
Reset-remove semantic on Map

### PR DESCRIPTION
**Pull request by russelldb**
With a shared causal context for the map and the CRDTs it embeds, removing a field, and then re-adding it has the effect of removing all the additions to that field when merging with a concurrent update to the field. For example:

```
Add 'X' to set in field F at A
Merge with B
Add 'Y' to set in field F at B
remove field F at A
Add 'Z' to field F at A
merge A and B
```

The set contains only 'Y' and 'Z', since the remove of the field at A containing 'X' constitutes a remove of 'X'.

However, if the field was _not_ re-added by A the semantic was different, and the result would contain 'X, Y, Z'. This seemed counter-intuitive and confusing. These changes add a temporary tombstone to present fields, so that the removal of a field effects a removal of all dots seen by the field at removal time. But you don't have to re-add the field to see this effect. Neat.

One intuitive way to think of it as like an offline sync of a file system like Dropbox. With this semantic, removing a directory on one machine, and adding a file to that directory on another, merges to the directory being present, with only the added file in it.

There is some oddness with counters, though.

I added a new counter type, `riak_dt_emcntr` (see it's docs for more details) that stores a `dot` as latest event against each actors increments and decrements. This is so a field removal can reset the counter value. However, we don't store a dot per increment (that would be crazy inefficient). Counters work will in this scenario

```
A increments counter at field F by 10
B merges A
A removes field F
B increments counter at field F by 5
A and B merge
```

Counter is at 5 (A's dot for the increment is dropped)

However there is an anomaly

```
A increments counter at field F by 10
B merges A
A increments counter by 5
B removes field F
A and B merge
```

The counter is at 15.

Now, technically, A's first increment _should_ be dropped (should it??) and only the second stand, but that does not happen, as we don't keep a dot per increment. This, too, can be spun in the filesystem sense. If you updated a file and concurrently removed it, you wouldn't expect only the diff to survive, but the whole file.

Personally I'm happy enough with the semantic, to enable reset-remove on counters requires a dot per increment (a counter as set of increments) and that is not worth the overhead.

Happy to rebase once the full round of reviewing is done.

OOOOOOOH, And I removed `add` from the Map API. Problem?

### CODE REVIEW

**Comment by russelldb** [comment]
This is kind of WIP, in that there is still W to do. But I thought it prudent to get eyes on early.

Still todo:-
- [x] docs for map (see todo in `riak_dt_map`)
- [x] riak_test for semantic
- [x] remove copypasta test from `riak_dt_emcntr`
- [x] riak_kv support of `riak_dt_emcntr`
- [ ]  riak_kv removal of map `add` (or re-add `add` to map with better semantic)

File: src/riak_dt_map.erl [hunk 1]
```diff
@@ -293,16 +339,118 @@ defer_remove(Clock, Ctx, Field, Deferred) ->
                                 Deferred)
     end.
 
-%% @doc merge two `map()'s.  and then a pairwise merge on all values
-%% in the value list.  This is the LUB function.
+%% @doc merge two `map()'s.
 -spec merge(map(), map()) -> map().
+merge(Map, Map) ->
+    Map;
+%% @TODO is there a way to optimise this, based on clocks maybe?
```
Comment (seancribbs) [inline src/riak_dt_map.erl:444]: You could short-circuit the case where one clock dominates the other, returning only the dominant one. Right?

Comment (russelldb) [inline src/riak_dt_map.erl:444]: That requires more analytical thinking than I'm capable of right now (thinking embedded maps more than the container.) Maybe…?

Comment (seancribbs) [inline src/riak_dt_map.erl:444]: Oh, right. I'll ponder it a while too, maybe we can figure it out tomorrow.

Comment (seancribbs) [inline src/riak_dt_map.erl:444]: Well, tomorrow came and went and we didn't come to any conclusion. That said, I can't see why it isn't safe to ignore one map entirely if its clock is dominated (or even just descended). My intuition says that these no-op merges would be mostly at the get_fsm where replicas are already in-sync, and so there would still be value there, by way of reduced latency.

**Comment by seancribbs** [comment]
Can you rebase or merge develop into this so it will merge cleanly (and make borshop happy)?

**Comment by russelldb** [comment]
After the review I can.

File: src/riak_dt_emcntr.erl [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+                {P::non_neg_integer(), N::non_neg_integer()}.
+op(Op, error) ->
+    op(Op, {0, 0});
+op(Op, {ok, {_Evt, P, N}}) ->
+    op(Op, {P, N});
+op(increment, {P, N}) ->
+    {P+1, N};
+op(decrement, {P, N}) ->
+    {P, N+1};
+op({increment, By}, {P, N}) ->
```
Comment (seancribbs) [inline src/riak_dt_emcntr.erl]: Nitpick: This doesn't handle the subtly-evil user who increments or decrements by a negative amount.

File: src/riak_dt_emcntr.erl [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+
+init_state() ->
+    0.
+
+gen_op() ->
+    oneof([increment,
+           {increment, nat()},
+           decrement,
+           {decrement, nat()},
+           {increment, ?LET(X, nat(), -X)}
```
Comment (seancribbs) [inline src/riak_dt_emcntr.erl]: Well, ignore my previous comment about `op/2`, it looks like negative increments are handled in the quickcheck property.

Comment (seancribbs) [inline src/riak_dt_emcntr.erl]: But, I still don't see how negative increments are handled properly. This doesn't have the EQC test across it by itself.

File: src/riak_dt_map.erl [hunk 4]
```diff
@@ -70,37 +84,46 @@
...
 
--type crdt()  ::  riak_dt_pncounter:pncounter() | riak_dt_od_flag:od_flag() |
+-type crdt()  ::  riak_dt_emcntr:emcntr() | riak_dt_od_flag:od_flag() |
                   riak_dt_lwwreg:lwwreg() |
                   riak_dt_orswot:orswot() |
                   riak_dt_map:map().
 
 -type map_op() :: {update, [map_field_update() | map_field_op()]}.
 
--type map_field_op() :: {add, field()} | {remove, field()}.
```
Comment (seancribbs) [inline src/riak_dt_map.erl:101]: Reminder to remove this op from `riak_pb` and `riak_kv_crdt_json`.

File: src/riak_dt_map.erl [hunk 5]
```diff
@@ -244,42 +267,65 @@ apply_ops([{add, {_Name, Mod}=Field} | Rest], Dot, {Clock, Values, Deferred}, Ct
...
+ctx_rem_field({_, Type}, {ok, {CRDTs, TS0}}, Ctx, MapClock) ->
+    %% Drop dominated fields, and update the tombstone.
+    %%
+    %% If the context is removing a field at dot {a, 1} and the
+    %% current field is {a, 2}, the tombstone ensures that all events
+    %% from {a, 1} are removed from the crdt value. If the ctx remove
+    %% is at {a, 3} and the current field is at {a, 2} then we need to
+    %% remove only events upto {a, 2}. The glb clock enables that.
+    %%
+    %% @TODO document how this doesn't really work for counters.
```
Comment (seancribbs) [inline src/riak_dt_map.erl]: Does the narrative in the PR description cover this TODO?

**Comment by seancribbs** [comment]
Is it ok that the `map_eqc` tends to have `0` as the most common actor count? Ran it for 6 minutes and got this output:

```
19.94% 0
11.44% 4
8.55% 3
8.31% 2
8.15% 6
[...]
```

File: src/riak_dt_map.erl (hunk 5 above)
Comment (russelldb) [inline src/riak_dt_map.erl]: Yes, does that count as "document"? If so, will happily cut that `@TODO` from src

**Comment by russelldb** [comment]
@seancribbs good question, I need to look at that some more. I mean, that's a high percentage, I think I need further metrics to see what size the map is, 'cos I suspect that there is still an issue with the state for this test. Luckily I have some expert eyes tomorrow.

File: src/riak_dt_map.erl (hunk 5 above)
Comment (seancribbs) [inline src/riak_dt_map.erl]: Yeah, maybe we just need a separate TODO for internal and external documentation.

File: src/riak_dt_map.erl (hunk 1 above)
Comment (seancribbs) [inline src/riak_dt_map.erl:444]: @russelldb On second thought, could deferred removes happen without incrementing the clock? If so, then my suggestion is moot.

Comment (russelldb) [inline src/riak_dt_map.erl:444]: Going to let eqc tell me. Deferreds are applied in merge, which merges the clocks.

Comment (russelldb) [inline src/riak_dt_map.erl:444]: OK, deferring this TODO for another branch. Want to get this semantic into the next Beta, optimisation comes after.

**Comment by seancribbs** [comment]
@russelldb I've rebased your branch, cleaning up a duplicate commit and grouping similar work. To get your local copy working:
1. `git stash` any unsaved changes you have
2. `git fetch origin` to get the latest head of the branch
3. `git reset --hard origin/feature-bug/rdb/removes-remove` to move your local branch to the rebased one
4. `git stash apply` to reapply your local changes

Ping me if you have trouble with this.

**Comment by russelldb** [comment]
OK. Addressed review comments and pushed. Just going to remove `add` from kv api, then ready for final review.

**Comment by russelldb** [comment]
Removed "add" from PB and HTTP APIs https://github.com/basho/riak_pb/pull/93, https://github.com/basho/riak_kv/pull/951

**Comment by seancribbs** [comment]
:+1: ffce004

**Comment by seancribbs** [comment]
@borshop merge

**Pull request merged**